File: tests/unit/expressions/agraph/test_simplification.py
"""Tests for bingo.expressions.agraph.simplification"""

import numpy as np
import pytest

from bingo.expressions.agraph.operators import (
    VARIABLE,
    CONSTANT,
    INTEGER,
    ADDITION,
    MULTIPLICATION,
    SIN,
)
from bingo.expressions.agraph.simplification import (
    get_utilized_commands,
    reduce,
)


class TestGetUtilizedCommands:
    def test_single_terminal(self):
        stack = np.array([[VARIABLE, 0, 0]], dtype=np.uint8)
        util = get_utilized_commands(stack)
        assert util == bytearray([1])

    def test_unused_command(self):
        # Row 0: X0, Row 1: X1 (unused), Row 2: sin(X0)
        stack = np.array(
            [
                [VARIABLE, 0, 0],
                [VARIABLE, 1, 1],
                [SIN, 0, 0],
            ],
            dtype=np.uint8,
        )
        util = get_utilized_commands(stack)
        assert util == bytearray([1, 0, 1])

    def test_binary_op_marks_both_children(self):
        stack = np.array(
            [
                [VARIABLE, 0, 0],
                [CONSTANT, 0, 0],
                [ADDITION, 0, 1],
            ],
            dtype=np.uint8,
        )
        util = get_utilized_commands(stack)
        assert util == bytearray([1, 1, 1])

    def test_chain_of_dependencies(self):
        # X0, X1, X0+X1, sin(X0+X1)
        stack = np.array(
            [
                [VARIABLE, 0, 0],
                [VARIABLE, 1, 1],
                [ADDITION, 0, 1],
                [SIN, 2, 2],
            ],
            dtype=np.uint8,
        )
        util = get_utilized_commands(stack)
        assert util == bytearray([1, 1, 1, 1])


class TestReduce:
    def test_no_reduction_needed(self):
        # All rows utilized; constants and integers pass through unchanged.
        stack = np.array(
            [
                [VARIABLE, 0, 0],
                [CONSTANT, 0, 0],
                [ADDITION, 0, 1],
            ],
            dtype=np.uint8,
        )
        reduced, consts, ints = reduce(stack, (3.14,), ())
        np.testing.assert_array_equal(
            reduced,
            np.array(
                [[VARIABLE, 0, 0], [CONSTANT, 0, 0], [ADDITION, 0, 1]],
                dtype=np.uint8,
            ),
        )
        assert consts == (3.14,)
        assert ints == ()

    def test_removes_unused_row(self):
        # Row 0: X0, Row 1: X1 (unused), Row 2: sin(X0)
        stack = np.array(
            [
                [VARIABLE, 0, 0],
                [VARIABLE, 1, 1],
                [SIN, 0, 0],
            ],
            dtype=np.uint8,
        )
        reduced, consts, ints = reduce(stack, (), ())
        expected = np.array(
            [[VARIABLE, 0, 0], [SIN, 0, 0]],
            dtype=np.uint8,
        )
        np.testing.assert_array_equal(reduced, expected)
        assert consts == ()
        assert ints == ()

    def test_remaps_operator_references(self):
        # Row 0: X0, Row 1: X1 (unused), Row 2: X1, Row 3: X0 + X1
        stack = np.array(
            [
                [VARIABLE, 0, 0],
                [VARIABLE, 1, 1],  # unused
                [VARIABLE, 1, 1],
                [ADDITION, 0, 2],
            ],
            dtype=np.uint8,
        )
        reduced, consts, ints = reduce(stack, (), ())
        expected = np.array(
            [
                [VARIABLE, 0, 0],
                [VARIABLE, 1, 1],
                [ADDITION, 0, 1],
            ],
            dtype=np.uint8,
        )
        np.testing.assert_array_equal(reduced, expected)
        assert consts == ()
        assert ints == ()

    def test_renumbers_constant_indices(self):
        # Row 0: C0, Row 1: C1 (unused), Row 2: C2, Row 3: C0 * C2
        stack = np.array(
            [
                [CONSTANT, 0, 0],
                [CONSTANT, 1, 1],  # unused
                [CONSTANT, 2, 2],
                [MULTIPLICATION, 0, 2],
            ],
            dtype=np.uint8,
        )
        raw_consts = (1.0, 99.0, 2.0)
        reduced, consts, ints = reduce(stack, raw_consts, ())
        # C1 is dropped; C0 -> index 0, C2 -> index 1
        expected = np.array(
            [
                [CONSTANT, 0, 0],
                [CONSTANT, 1, 1],
                [MULTIPLICATION, 0, 1],
            ],
            dtype=np.uint8,
        )
        np.testing.assert_array_equal(reduced, expected)
        assert consts == (1.0, 2.0)
        assert ints == ()

    def test_renumbers_integer_indices(self):
        # Row 0: I0, Row 1: I1 (unused), Row 2: I2, Row 3: I0 + I2
        stack = np.array(
            [
                [INTEGER, 0, 0],
                [INTEGER, 1, 1],  # unused
                [INTEGER, 2, 2],
                [ADDITION, 0, 2],
            ],
            dtype=np.uint8,
        )
        raw_ints = (10, 99, 20)
        reduced, consts, ints = reduce(stack, (), raw_ints)
        expected = np.array(
            [
                [INTEGER, 0, 0],
                [INTEGER, 1, 1],
                [ADDITION, 0, 1],
            ],
            dtype=np.uint8,
        )
        np.testing.assert_array_equal(reduced, expected)
        assert consts == ()
        assert ints == (10, 20)

    def test_preserves_dtype(self):
        stack = np.array([[VARIABLE, 0, 0]], dtype=np.uint8)
        reduced, _, _ = reduce(stack, (), ())
        assert reduced.dtype == np.uint8

    def test_empty_stack(self):
        stack = np.empty((0, 3), dtype=np.uint8)
        reduced, consts, ints = reduce(stack, (), ())
        assert reduced.shape == (0, 3)
        assert consts == ()
        assert ints == ()

File: bingo/expressions/agraph/cppagraph/tests/test_simplification.cpp
/**
 * @file test_simplification.cpp
 * @brief Google Test suite for cppagraph stack reduction.
 *
 * Mirrors the test scenarios from
 * tests/unit/expressions/agraph/pyagraph/test_simplification.py.
 */

#include <gtest/gtest.h>

#include <vector>

#include "cppagraph/operators.h"
#include "cppagraph/simplification.h"

using namespace cppagraph;

// ================================================================
//  Helpers
// ================================================================

static StackMatrix make_stack(
        std::initializer_list<std::array<uint8_t, 3>> rows) {
    StackMatrix s(static_cast<Eigen::Index>(rows.size()), 3);
    int i = 0;
    for (auto& r : rows) {
        s(i, 0) = r[0];
        s(i, 1) = r[1];
        s(i, 2) = r[2];
        ++i;
    }
    return s;
}

// Operator shorthands
static constexpr uint8_t VAR  = static_cast<uint8_t>(Op::VARIABLE);
static constexpr uint8_t CON  = static_cast<uint8_t>(Op::CONSTANT);
static constexpr uint8_t INT  = static_cast<uint8_t>(Op::INTEGER);
static constexpr uint8_t ADD  = static_cast<uint8_t>(Op::ADDITION);
static constexpr uint8_t MUL  = static_cast<uint8_t>(Op::MULTIPLICATION);
static constexpr uint8_t SSIN = static_cast<uint8_t>(Op::SIN);

// ================================================================
//  get_utilized_commands tests
// ================================================================

TEST(GetUtilizedCommands, SingleTerminal) {
    auto stack = make_stack({{VAR, 0, 0}});
    auto util = get_utilized_commands(stack);
    ASSERT_EQ(util.size(), 1u);
    EXPECT_TRUE(util[0]);
}

TEST(GetUtilizedCommands, UnusedCommand) {
    // Row 0: X0, Row 1: X1 (unused), Row 2: sin(X0)
    auto stack = make_stack({
        {VAR, 0, 0},
        {VAR, 1, 1},
        {SSIN, 0, 0},
    });
    auto util = get_utilized_commands(stack);
    EXPECT_TRUE(util[0]);
    EXPECT_FALSE(util[1]);
    EXPECT_TRUE(util[2]);
}

TEST(GetUtilizedCommands, BinaryOpMarksBothChildren) {
    auto stack = make_stack({
        {VAR, 0, 0},
        {CON, 0, 0},
        {ADD, 0, 1},
    });
    auto util = get_utilized_commands(stack);
    EXPECT_TRUE(util[0]);
    EXPECT_TRUE(util[1]);
    EXPECT_TRUE(util[2]);
}

TEST(GetUtilizedCommands, ChainOfDependencies) {
    // X0, X1, X0+X1, sin(X0+X1)
    auto stack = make_stack({
        {VAR, 0, 0},
        {VAR, 1, 1},
        {ADD, 0, 1},
        {SSIN, 2, 2},
    });
    auto util = get_utilized_commands(stack);
    for (int i = 0; i < 4; ++i)
        EXPECT_TRUE(util[i]);
}

TEST(GetUtilizedCommands, EmptyStack) {
    StackMatrix stack(0, 3);
    auto util = get_utilized_commands(stack);
    EXPECT_TRUE(util.empty());
}

// ================================================================
//  reduce tests
// ================================================================

TEST(Reduce, NoReductionNeeded) {
    auto stack = make_stack({
        {VAR, 0, 0},
        {CON, 0, 0},
        {ADD, 0, 1},
    });
    auto r = reduce(stack, {3.14}, {});

    ASSERT_EQ(r.stack.rows(), 3);
    EXPECT_EQ(r.stack(0, 0), VAR);
    EXPECT_EQ(r.stack(0, 1), 0);
    EXPECT_EQ(r.stack(1, 0), CON);
    EXPECT_EQ(r.stack(1, 1), 0);
    EXPECT_EQ(r.stack(2, 0), ADD);
    EXPECT_EQ(r.stack(2, 1), 0);
    EXPECT_EQ(r.stack(2, 2), 1);
    ASSERT_EQ(r.constants.size(), 1u);
    EXPECT_DOUBLE_EQ(r.constants[0], 3.14);
    EXPECT_TRUE(r.integers.empty());
}

TEST(Reduce, RemovesUnusedRow) {
    // Row 0: X0, Row 1: X1 (unused), Row 2: sin(X0)
    auto stack = make_stack({
        {VAR, 0, 0},
        {VAR, 1, 1},
        {SSIN, 0, 0},
    });
    auto r = reduce(stack, {}, {});

    ASSERT_EQ(r.stack.rows(), 2);
    EXPECT_EQ(r.stack(0, 0), VAR);
    EXPECT_EQ(r.stack(0, 1), 0);
    EXPECT_EQ(r.stack(1, 0), SSIN);
    EXPECT_EQ(r.stack(1, 1), 0);
}

TEST(Reduce, RemapsOperatorReferences) {
    // Row 0: X0, Row 1: X1 (unused), Row 2: X1, Row 3: X0 + X1
    auto stack = make_stack({
        {VAR, 0, 0},
        {VAR, 1, 1},   // unused
        {VAR, 1, 1},
        {ADD, 0, 2},
    });
    auto r = reduce(stack, {}, {});

    ASSERT_EQ(r.stack.rows(), 3);
    EXPECT_EQ(r.stack(0, 0), VAR);
    EXPECT_EQ(r.stack(0, 1), 0);
    EXPECT_EQ(r.stack(1, 0), VAR);
    EXPECT_EQ(r.stack(1, 1), 1);
    EXPECT_EQ(r.stack(2, 0), ADD);
    EXPECT_EQ(r.stack(2, 1), 0);
    EXPECT_EQ(r.stack(2, 2), 1);
}

TEST(Reduce, RenumbersConstantIndices) {
    // C0, C1 (unused), C2, C0*C2
    auto stack = make_stack({
        {CON, 0, 0},
        {CON, 1, 1},   // unused
        {CON, 2, 2},
        {MUL, 0, 2},
    });
    auto r = reduce(stack, {1.0, 99.0, 2.0}, {});

    ASSERT_EQ(r.stack.rows(), 3);
    EXPECT_EQ(r.stack(0, 0), CON);
    EXPECT_EQ(r.stack(0, 1), 0);
    EXPECT_EQ(r.stack(1, 0), CON);
    EXPECT_EQ(r.stack(1, 1), 1);
    EXPECT_EQ(r.stack(2, 0), MUL);
    EXPECT_EQ(r.stack(2, 1), 0);
    EXPECT_EQ(r.stack(2, 2), 1);
    ASSERT_EQ(r.constants.size(), 2u);
    EXPECT_DOUBLE_EQ(r.constants[0], 1.0);
    EXPECT_DOUBLE_EQ(r.constants[1], 2.0);
}

TEST(Reduce, RenumbersIntegerIndices) {
    // I0, I1 (unused), I2, I0+I2
    auto stack = make_stack({
        {INT, 0, 0},
        {INT, 1, 1},   // unused
        {INT, 2, 2},
        {ADD, 0, 2},
    });
    auto r = reduce(stack, {}, {10, 99, 20});

    ASSERT_EQ(r.stack.rows(), 3);
    EXPECT_EQ(r.stack(0, 0), INT);
    EXPECT_EQ(r.stack(0, 1), 0);
    EXPECT_EQ(r.stack(1, 0), INT);
    EXPECT_EQ(r.stack(1, 1), 1);
    EXPECT_EQ(r.stack(2, 0), ADD);
    EXPECT_EQ(r.stack(2, 1), 0);
    EXPECT_EQ(r.stack(2, 2), 1);
    ASSERT_EQ(r.integers.size(), 2u);
    EXPECT_EQ(r.integers[0], 10);
    EXPECT_EQ(r.integers[1], 20);
}

TEST(Reduce, EmptyStack) {
    StackMatrix stack(0, 3);
    auto r = reduce(stack, {}, {});
    EXPECT_EQ(r.stack.rows(), 0);
    EXPECT_TRUE(r.constants.empty());
    EXPECT_TRUE(r.integers.empty());
    EXPECT_TRUE(r.constant_mapping.empty());
}

// ================================================================
//  Constant mapping tests
// ================================================================

TEST(ReduceConstantMapping, IdentityMappingSingleConstant) {
    auto stack = make_stack({
        {VAR, 0, 0},
        {CON, 0, 0},
        {ADD, 0, 1},
    });
    auto r = reduce(stack, {3.14}, {});
    ASSERT_EQ(r.constant_mapping.size(), 1u);
    EXPECT_EQ(r.constant_mapping[0], 0);
}

TEST(ReduceConstantMapping, SkipsUnusedConstant) {
    // C0 used, C1 unused, C2 used → mapping is {0, 2}
    auto stack = make_stack({
        {CON, 0, 0},
        {CON, 1, 1},   // unused
        {CON, 2, 2},
        {MUL, 0, 2},
    });
    auto r = reduce(stack, {1.0, 99.0, 2.0}, {});
    ASSERT_EQ(r.constant_mapping.size(), 2u);
    EXPECT_EQ(r.constant_mapping[0], 0);
    EXPECT_EQ(r.constant_mapping[1], 2);
}

TEST(ReduceConstantMapping, PreservesOrder) {
    // C2 first, then C0 → mapping is {2, 0}
    auto stack = make_stack({
        {CON, 2, 2},
        {CON, 0, 0},
        {ADD, 0, 1},
    });
    auto r = reduce(stack, {10.0, 20.0, 30.0}, {});
    ASSERT_EQ(r.constant_mapping.size(), 2u);
    EXPECT_EQ(r.constant_mapping[0], 2);
    EXPECT_EQ(r.constant_mapping[1], 0);
}

TEST(ReduceConstantMapping, EmptyMappingNoConstants) {
    auto stack = make_stack({
        {VAR, 0, 0},
        {SSIN, 0, 0},
    });
    auto r = reduce(stack, {}, {});
    EXPECT_TRUE(r.constant_mapping.empty());
}

TEST(ReduceConstantMapping, MappingValuesIndexIntoRawConstants) {
    auto stack = make_stack({
        {CON, 0, 0},
        {CON, 1, 1},   // unused
        {CON, 2, 2},
        {MUL, 0, 2},
    });
    std::vector<double> raw_consts = {10.0, 99.0, 20.0};
    auto r = reduce(stack, raw_consts, {});
    for (size_t i = 0; i < r.constant_mapping.size(); ++i) {
        EXPECT_DOUBLE_EQ(r.constants[i], raw_consts[r.constant_mapping[i]]);
    }
}
